Speedup segmented sort with large nsort

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77188

Follow up to gh-77100

Instead of calling `at::arange`, I repurpose the existing kernels to
achieve the same effect. I've also fixed the 2d case bug where
the pointer was advanced by `n` which equals `nsegment * nsort` after
only processing `nsort` elements.

peterbell10 committed:
Speedup segmented sort with large nsort

Follow up to gh-77100

Instead of calling `at::arange`, I repurpose the existing kernels to
achieve the same effect. I've also fixed the 2d case bug where
the pointer was advanced by `n` which equals `nsegment * nsort` after
only processing `nsort` elements.

[ghstack-poisoned]

ngimel commented:
I reverted that PR because of nsort bug already :-)

peterbell10 commented:
I already rebased to after the revert before submitting the PR.

ngimel commented:
Can you also please add a test that I added to the reverted PR?

ngimel commented on aten/src/ATen/native/cuda/Sort.cu:
```diff
@@ -241,13 +243,38 @@ __global__ void fill_index_and_segment_kernel(
+  at::cuda::detail::IntDivider<uint32_t> nsort_divider(nsort);
+
+  for (auto i: c10::irange(nsegments)){
+    fill_reverse_indices_kernel<<<grid, block, 0, stream>>>(
```
oh, my mistake - we don't need to refill indices every time as cub doesn't modify them.

ngimel reviewed (approved):
This looks good, but I can't merge now as github is rate limited and CI is not running

peterbell10 committed:
Update on "Speedup segmented sort with large nsort"


Follow up to gh-77100

Instead of calling `at::arange`, I repurpose the existing kernels to
achieve the same effect. I've also fixed the 2d case bug where
the pointer was advanced by `n` which equals `nsegment * nsort` after
only processing `nsort` elements.

[ghstack-poisoned]

peterbell10 committed:
Fix lint on "Speedup segmented sort with large nsort"


Follow up to gh-77100

Instead of calling `at::arange`, I repurpose the existing kernels to
achieve the same effect. I've also fixed the 2d case bug where
the pointer was advanced by `n` which equals `nsegment * nsort` after
only processing `nsort` elements.

[ghstack-poisoned]

peterbell10 committed:
Rebase to get ROCm CI signal on "Speedup segmented sort with large nsort"


Follow up to gh-77100

Instead of calling `at::arange`, I repurpose the existing kernels to
achieve the same effect. I've also fixed the 2d case bug where
the pointer was advanced by `n` which equals `nsegment * nsort` after
only processing `nsort` elements.

[ghstack-poisoned]
